From f06e5c0c8e055f33f13f76d586338714c6a50d55 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Sat, 25 Apr 2020 07:43:57 +0300
Subject: [PATCH] [SW-0] uart: add getc()/tstc()

Now it is possible to read characters from uart
It may be useful to get some indication from user e.g.
Esc character is used to log OAM index
It is used by BMC for knowing which tty belongs to which module id

Change-Id: I33c794def41e5cc6f8efe59dd8ccbf56a0555107
---

diff --git a/common/uart.c b/common/uart.c
index 8e12eb8..5a8990b 100644
--- a/common/uart.c
+++ b/common/uart.c
@@ -12,6 +12,7 @@
 #include "uart.h"
 
 #define UART_SR_TXFULL			(1 << 0) /* TX Buffer Full */
+#define UART_SR_RXFULL			(1 << 1) /* RX Buffer Full */
 #define STR_UINT_SZ			8
 #define BITS_IN_SINGLE_CHAR		4
 
@@ -22,6 +23,22 @@
 	regmap_writel(0x3, mmPSOC_UART_0_CTRL);
 }
 
+int preboot_uart_tstc(void)
+{
+	if (!(regmap_readl(mmPSOC_UART_0_STATE) & UART_SR_RXFULL))
+		return 0;
+
+	return 1;
+}
+
+int preboot_uart_getc(void)
+{
+	if (!preboot_uart_tstc())
+		return -1;
+
+	return regmap_readl(mmPSOC_UART_0_DATA);
+}
+
 void preboot_putc_raw(const char ch)
 {
 	while (regmap_readl(mmPSOC_UART_0_STATE) & UART_SR_TXFULL)
diff --git a/common/uart.h b/common/uart.h
index 7627a3e..2bc45f0 100644
--- a/common/uart.h
+++ b/common/uart.h
@@ -20,5 +20,7 @@
 void preboot_puts_hex_raw(const char *log, unsigned int val, bool pad);
 void preboot_puts_hex(const char *log, unsigned int val, bool pad);
 void preboot_puts_char(const char *str, unsigned char val);
+int preboot_uart_tstc(void);
+int preboot_uart_getc(void);
 
 #endif /* _HL_UART_H_ */
